planner/core/rule_max_min_eliminate.go

```diff
@@ -18,71 +18,188 @@ import (
+// because we have restricted the subPlan in `checkColCanUseIndex`.
+func (a *maxMinEliminator) cloneSubPlans(plan LogicalPlan) LogicalPlan {
+	switch p := plan.(type) {
+	case *LogicalSelection:
+		conditions := make([]expression.Expression, 0, len(p.Conditions))
+		for _, cond := range p.Conditions {
+			conditions = append(conditions, cond.Clone())
+		}
+		sel := LogicalSelection{Conditions: conditions}.Init(p.ctx)
+		sel.SetChildren(a.cloneSubPlans(p.children[0]))
+		return sel
+	case *DataSource:
+		// This is a shallow clone which may not be safe.
+		newDs := *p
+		newDs.self = &newDs
+		return &newDs
+	}
+	// This won't happen, because we have checked the subtree.
+	return nil
+}
 
-			// Add Sort and Limit operators.
-			// For max function, the sort order should be desc.
-			desc := f.Name == ast.AggFuncMax
-			// Compose Sort operator.
-			sort := LogicalSort{}.Init(ctx)
-			sort.ByItems = append(sort.ByItems, &ByItems{f.Args[0], desc})
-			sort.SetChildren(child)
-			child = sort
+// splitAggFuncAndCheckIndices splits the agg to multiple aggs and check whether each agg needs a sort
+// after the transformation. For example, we firstly split the sql: `select max(a), min(a), max(b) from t` ->
+// `select max(a) from t` + `select min(a) from t` + `select max(b) from t`.
+// Then we check whether `a` and `b` have indices. If any of the used column has no index, we cannot eliminate
+// this aggregation.
+func (a *maxMinEliminator) splitAggFuncAndCheckIndices(agg *LogicalAggregation) (aggs []*LogicalAggregation, canEliminate bool) {
+	aggs = make([]*LogicalAggregation, 0, len(agg.AggFuncs))
+	for _, f := range agg.AggFuncs {
+		// We must make sure the args of max/min is a simple single column.
+		col, ok := f.Args[0].(*expression.Column)
+		if !ok {
+			return nil, false
 		}
+		if !a.checkColCanUseIndex(agg.children[0], col) {
+			return nil, false
+		}
+	}
+	// we can split the aggregation only if all of the aggFuncs pass the check.
+	for i, f := range agg.AggFuncs {
+		newAgg := LogicalAggregation{AggFuncs: []*aggregation.AggFuncDesc{f}}.Init(agg.ctx)
+		newAgg.SetChildren(a.cloneSubPlans(agg.children[0]))
+		newAgg.schema = expression.NewSchema(agg.schema.Columns[i])
+		aggs = append(aggs, newAgg)
+	}
+	return aggs, true
+}
 
-		// Compose Limit operator.
-		li := LogicalLimit{Count: 1}.Init(ctx)
-		li.SetChildren(child)
+// Try to convert a single max/min to Limit+Sort operators.
+func (a *maxMinEliminator) eliminateSingleMaxMin(agg *LogicalAggregation) *LogicalAggregation {
+	f := agg.AggFuncs[0]
+	child := agg.Children()[0]
+	ctx := agg.SCtx()
 
-		// If no data in the child, we need to return NULL instead of empty. This cannot be done by sort and limit themselves.
-		// Since now it's almost one row returned, a agg operator is okay to do this.
-		p.SetChildren(li)
-		return
+	// If there's no column in f.GetArgs()[0], we still need limit and read data from real table because the result should NULL if the below is empty.
```

Comment:
```suggestion
	// If there's no column in f.GetArgs()[0], we still need limit and read data from real table because the result should be NULL if the input is empty.
```